src/theory/strings/rewrites

```diff
@@ -422,6 +495,100 @@
   (str.contains (str.replace s t r) u)
   true)
 
+(define-cond-rule str-replace-dual-ctn-false ((s ?Seq) (t ?Seq) (r ?Seq) (u ?Seq))
+  (and (not (str.contains s t)) (not (str.contains s u)))
+  (str.contains s (str.replace t r u))
+  false)
+
+(define-rule str-replace-self-ctn-simp ((s ?Seq) (t ?Seq))
+  (str.contains s (str.replace t s t))
+  (str.contains s t))
+
+(define-cond-rule str-replace-emp-ctn-src ((s ?Seq) (t ?Seq) (emp ?Seq))
+  (= emp (@seq.empty_of_type (@type_of s)))
+  (str.contains s (str.replace emp s t))
+  (= emp (str.replace emp s t)))
+
+(define-cond-rule str-substr-char-start-eq-len ((x ?Seq) (n Int))
+  (>= 1 (str.len x))
+  (str.substr x n n)
+  (@seq.empty_of_type (@type_of x)))
+
+(define-cond-rule str-contains-repl-char ((x ?Seq) (y ?Seq) (z ?Seq) (w ?Seq))
+  (and (= (str.len w) 1) (not (str.contains w y)))
+  (str.contains (str.replace x y z) w)
+  (or (str.contains x w) (and (str.contains x y) (str.contains z w))))
+
+(define-cond-rule str-contains-repl-self-tgt-char ((x ?Seq) (y ?Seq) (w ?Seq))
+  (= (str.len w) 1)
+  (str.contains (str.replace x y x) w)
+  (str.contains x w))
+
+(define-rule str-contains-repl-self ((x ?Seq) (y ?Seq))
+  (str.contains (str.replace x y x) y)
+  (str.contains x y))
+
+(define-rule str-contains-repl-tgt ((x ?Seq) (y ?Seq) (z ?Seq))
+  (str.contains (str.replace x y z) z)
+  (or (str.contains x y) (str.contains x z)))
+
+(define-cond-rule str-repl-repl-len-id ((x ?Seq) (y ?Seq))
+  (>= (str.len y) (str.len x))
+  (str.replace x y x)
+  x)
+
+(define-cond-rule str-repl-repl-src-tgt-no-ctn ((x ?Seq) (y ?Seq) (z ?Seq) (w ?Seq))
```

Comment:
Sorry, but could you intuitively explain what this rule does. I don't quite get it.

Is it just that since z cannot contain w, that if z contains x, x cannot contain w?

Reply:
Yes, this one is quite tricky:

Essentially the main argument is that either:
(A) `x` does not contain `w` and hence  `(str.replace z x y)` is irrelevant.
(B) `x` contains `w` but `z` does not contain `x`. This is because if `x` contains `w` and `z` does not contain `w` by the premise, hence `z` does not contain `x`.

In either case `(str.replace z x y)` can be replaced by `z`.

Reply:
Perfect, thanks